File: backend/app/models/scraping.py
"""
Scraping API models.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, model_validator

from app.database.models.scraping_job import JobStatusEnum


class ScrapingJobCreate(BaseModel):
    """Request to create a scraping job."""
    company_id: str = Field(..., description="Company ID to scrape reviews for")
    source_id: str = Field(..., description="Review source ID (Reddit, Twitter, etc.)")
    review_count: Optional[int] = Field(None, ge=1, le=1000, description="Number of reviews to scrape/generate")
    max_cost: Optional[float] = Field(None, ge=0.01, le=1000.0, description="Maximum cost in credits")
    generation_mode: Optional[str] = Field(None, description="Mode: 'fake' for LLM generation, 'real' for actual scraping")

    class Config:
        json_schema_extra = {
            "example": {
                "company_id": "comp_123",
                "source_id": "source_reddit",
                "review_count": 100,
                "generation_mode": "fake"
            }
        }


class ScrapingJobResponse(BaseModel):
    """Scraping job response."""
    id: str
    company_id: str
    source_id: str
    user_id: str
    status: JobStatusEnum
    progress_percentage: float
    total_reviews_target: int
    reviews_fetched: int
    cost: float
    celery_task_id: Optional[str]
    error_message: Optional[str]
    created_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    # Human-readable names
    source_name: Optional[str] = None
    company_name: Optional[str] = None

    class Config:
        from_attributes = True


class CostEstimate(BaseModel):
    """Cost estimate for scraping."""
    source_name: str
    review_count: int
    cost: float
    cost_per_review: float

File: frontend/src/types/scraping.ts
export interface ScrapingSource {
  id: string;
  name: string;
  source_type: 'reddit' | 'twitter' | 'custom';
  cost_per_review: number;
  is_active: boolean;
  description?: string;
  icon?: string;
}

export interface ScrapingJob {
  id: string;
  company_id: string;
  source_id: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  progress_percentage: number;
  total_reviews_target: number;
  reviews_fetched: number;
  cost: number;
  started_at?: string;
  completed_at?: string;
  created_at?: string;
  error_message?: string;
  user_id: string;
  // Human-readable names
  source_name?: string;
  company_name?: string;
}

export interface StartScrapingJobRequest {
  company_id: string;
  source_id: string;
  total_reviews_target: number;
}

export interface ScrapingJobEstimate {
  estimated_cost: number;
  estimated_duration_minutes: number;
  source_name: string;
}

export interface Review {
  id: string;
  company_id: string;
  source_id: string;
  scraping_job_id?: string;
  content: string;
  author: string;
  sentiment_score: number;
  url?: string;
  metadata?: Record<string, any>;
  scraped_at: string;
  processed_at?: string;
}

export interface DataImport {
  id: string;
  company_id: string;
  user_id: string;
  file_path: string;
  import_type: 'csv' | 'json';
  status: 'pending' | 'processing' | 'completed' | 'failed';
  rows_imported: number;
  total_rows?: number;
  created_at: string;
  completed_at?: string;
  error_message?: string;
}

